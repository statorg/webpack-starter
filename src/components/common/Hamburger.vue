<template>
  <button
    class="hamburger"
    :class="classes"
    type="button"
    aria-label="Toggle menu"
    @click.prevent="toggle"
  >
    <span class="hamburger__inner">
      <span class="hamburger__bar" />
      <span class="hamburger__bar" />
    </span>
    <span class="hamburger__bar" />
  </button>
</template>

<script>
export default {
  name: 'Hamburger',
  model: {
    prop: 'active',
    event: 'change'
  },
  props: {
    active: Boolean
  },
  data () {
    return {
      isActive: this.active
    }
  },
  computed: {
    classes () {
      return {
        hamburger_active: this.isActive
      }
    }
  },
  watch: {
    active (value) {
      this.isActive = value
    },
    isActive (value) {
      this.$emit('change', value)
    }
  },
  methods: {
    toggle () {
      this.isActive = !this.isActive
    }
  }
}
</script>

<style lang="scss">
$hamburger-width: 26px;
$hamburger-height: 22px;
$hamburger-bar-height: 2px;
$hamburger-bar-radius: 0;
$hamburger-bar-offset: ($hamburger-height - $hamburger-bar-height) / 2;
$hamburger-transition: 0.3s easing(fast-in-fast-out);

.hamburger {
  width: $hamburger-width;
  height: $hamburger-height;
  background-color: transparent;
  position: relative;
  display: inline-block;
  vertical-align: middle;
  border: none;
  padding: 0;
  cursor: pointer;
  transform: translate3d(0, 0, 0);
  transition: $hamburger-transition;
  color: #222222;

  &__inner {
    position: absolute;
    top: 0;
    left: 50%;
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    transform: translateX(-50%);
    transition: inherit;
  }

  &_active &__inner {
    width: 110%;
  }

  &__bar {
    height: $hamburger-bar-height;
    background-color: currentColor;
    width: 100%;
    display: block;
    border-radius: $hamburger-bar-radius;
    transition: inherit;
    transform: translate3d(0, 0, 0);
  }

  & > &__bar {
    position: absolute;
    top: 0;
    bottom: 0;
    margin: auto 0;
  }

  &_active > &__bar {
    background-color: transparent;
  }

  &_active &__inner > &__bar {
    &:first-child {
      transform: translate3d(0, $hamburger-bar-offset, 0) rotate(45deg);
    }
    &:last-child {
      transform: translate3d(0, $hamburger-bar-offset * -1, 0) rotate(-45deg);
    }
  }
}

@include vue(v-hamburger) {
  @extend .hamburger;

  &:before,
  &:after {
    content: '';
    position: absolute;
    left: 0;
    width: 100%;
    height: $hamburger-bar-height;
    border-radius: $hamburger-bar-radius;
    background-color: currentColor;
  }
  &:before {
    top: 0;
    box-shadow: 0 $hamburger-bar-offset currentColor;
  }
  &:after {
    bottom: 0;
  }
}
</style>
