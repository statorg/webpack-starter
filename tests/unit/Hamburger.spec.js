import { shallowMount } from '@vue/test-utils'
import Hamburger from '@/components/common/Hamburger'

describe('Hamburger.vue', () => {
  it('should have hamburger_active class when active prop is passed as true', () => {
    const wrapper = shallowMount(Hamburger, {
      propsData: {
        active: true
      }
    })
    expect(wrapper.classes()).toContain('hamburger_active')
  })

  it('should emit change event when clicked', () => {
    const wrapper = shallowMount(Hamburger)
    wrapper.trigger('click')
    expect(wrapper.emitted('change')).toBeTruthy()
    expect(wrapper.emitted('change').length).toBe(1)
    expect(wrapper.emitted('change')[0]).toEqual([true])
  })

  it('should render and match snapshot', () => {
    const wrapper = shallowMount(Hamburger)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
