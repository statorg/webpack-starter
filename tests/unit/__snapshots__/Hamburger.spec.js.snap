// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger.vue should render and match snapshot 1`] = `<button type="button" aria-label="Toggle menu" class="hamburger"><span class="hamburger__inner"><span class="hamburger__bar"></span> <span class="hamburger__bar"></span></span> <span class="hamburger__bar"></span></button>`;
